File: app/src/main/java/com/example/wilfred/journalapp/database/EntryRepository.java
package com.example.wilfred.journalapp.database;

import android.app.Application;
import android.arch.lifecycle.LiveData;
import android.os.AsyncTask;

import java.util.List;

/**
 * Created by wilfred on 6/30/18.
 */

public class EntryRepository {
    private EntryDao mEntryDao;
    private LiveData<List<JournalEntry>> mEntries;

    public EntryRepository(Application application) {
        JournalDatabase database = JournalDatabase.getsInstance(application);
        mEntryDao = database.entryDao();
        mEntries = mEntryDao.getAllEntries();
    }

    //AsyncTask to retrieve all entries
    public LiveData<List<JournalEntry>> getEntries() {
        return mEntries;
    }

    public LiveData<JournalEntry> getEntryById(int id) {
        return mEntryDao.getEntryById(id);
    }

    public void insert(JournalEntry entry) {
        new insertAsyncTask(mEntryDao).execute(entry);
    }


    //AsyncTask to delete an entry
    void delete(JournalEntry entry) {
        new deleteAsyncTask(mEntryDao).execute(entry);
    }

    public void updateJournal(JournalEntry entry) {
        new updateAsyncTask(mEntryDao).execute(entry);

    }

    private static class insertAsyncTask extends AsyncTask<JournalEntry, Void, Void> {

        private EntryDao mAsyncTaskDao;

        insertAsyncTask(EntryDao dao) {
            mAsyncTaskDao = dao;
        }

        @Override
        protected Void doInBackground(final JournalEntry... params) {
            mAsyncTaskDao.insertEntry(params[0]);
            return null;
        }
    }

    //Update Async
    private static class updateAsyncTask extends AsyncTask<JournalEntry, Void, Void> {

        private EntryDao mAsyncDao;

        updateAsyncTask(EntryDao dao) {
            mAsyncDao = dao;
        }

        @Override
        protected Void doInBackground(final JournalEntry... journalEntries) {
            mAsyncDao.updateEntry(journalEntries[0]);
            return null;
        }
    }


    //Delete AsyncTask
    private static class deleteAsyncTask extends AsyncTask<JournalEntry, Void, Void> {

        private EntryDao mAsyncTaskDao;

        deleteAsyncTask(EntryDao dao) {
            mAsyncTaskDao = dao;
        }

        @Override
        protected Void doInBackground(final JournalEntry... params) {
            mAsyncTaskDao.deleteEntry(params[0]);
            return null;
        }
    }
}

File: app/src/main/java/com/example/wilfred/journalapp/database/EntryRepository.kt
package com.example.wilfred.journalapp.database

import android.app.Application
import androidx.lifecycle.LiveData
import android.os.AsyncTask

/**
 * Created by wilfred on 6/30/18.
 */

class EntryRepository(application: Application) {
    private val mEntryDao: EntryDao
    //AsyncTask to retrieve all entries
    var entries: LiveData<List<JournalEntry>>

    init {
        val database = JournalDatabase.getsInstance(application)
        mEntryDao = database!!.entryDao()
        entries = mEntryDao.allEntries
    }

    fun getEntryById(id: Int): LiveData<JournalEntry> {
        return mEntryDao.getEntryById(id)
    }

    fun insert(entry: JournalEntry) {
        insertAsyncTask(mEntryDao).execute(entry)
    }


    //AsyncTask to delete an entry
    internal fun delete(entry: JournalEntry) {
        deleteAsyncTask(mEntryDao).execute(entry)
    }

    fun updateJournal(entry: JournalEntry) {
        updateAsyncTask(mEntryDao).execute(entry)

    }

    private class insertAsyncTask internal constructor(private val mAsyncTaskDao: EntryDao) : AsyncTask<JournalEntry, Void, Void>() {

        override fun doInBackground(vararg params: JournalEntry): Void? {
            mAsyncTaskDao.insertEntry(params[0])
            return null
        }
    }

    //Update Async
    private class updateAsyncTask internal constructor(private val mAsyncDao: EntryDao) : AsyncTask<JournalEntry, Void, Void>() {

        override fun doInBackground(vararg journalEntries: JournalEntry): Void? {
            mAsyncDao.updateEntry(journalEntries[0])
            return null
        }
    }


    //Delete AsyncTask
    private class deleteAsyncTask internal constructor(private val mAsyncTaskDao: EntryDao) : AsyncTask<JournalEntry, Void, Void>() {

        override fun doInBackground(vararg params: JournalEntry): Void? {
            mAsyncTaskDao.deleteEntry(params[0])
            return null
        }
    }
}
